plugins/uefi/fu-plugin-uefi.c

```diff
@@ -135,6 +137,202 @@ fu_plugin_get_results (FuPlugin *plugin, FuDevice *device, GError **error)
+			     "UEFI firmware update failed: %s",
+			     str->str);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static GBytes *
+fu_plugin_uefi_get_splash_data (guint width, guint height, GError **error)
+{
+	const gchar * const *langs = g_get_language_names ();
+	const gchar *localedir = LOCALEDIR;
+	const gsize chunk_size = 1024 * 1024;
+	gsize buf_idx = 0;
+	gsize buf_sz = chunk_size;
+	gssize len;
+	g_autofree gchar *basename = NULL;
+	g_autofree guint8 *buf = NULL;
+	g_autoptr(GBytes) compressed_data = NULL;
+	g_autoptr(GConverter) conv = NULL;
+	g_autoptr(GInputStream) stream_compressed = NULL;
+	g_autoptr(GInputStream) stream_raw = NULL;
+
+	/* ensure this is sane */
+	if (!g_str_has_prefix (localedir, "/"))
+		localedir = "/usr/share/locale";
+
+	/* find the closest locale match */
+	basename = g_strdup_printf ("fwupd-%u-%u.bmp.gz", width, height);
+	for (guint i = 0; langs[i] != NULL; i++) {
+		g_autofree gchar *fn = NULL;
+		if (g_str_has_suffix (langs[i], ".UTF-8"))
+			continue;
+		fn = g_build_filename (localedir, langs[i],
+				       "LC_IMAGES", basename, NULL);
+		if (g_file_test (fn, G_FILE_TEST_EXISTS)) {
+			compressed_data = fu_common_get_contents_bytes (fn, error);
+			if (compressed_data == NULL)
+				return NULL;
+			break;
+		}
+		g_debug ("no %s found", fn);
+	}
+
+	/* we found nothing, even LANG=en and LANG=C */
+	if (compressed_data == NULL) {
+		g_autofree gchar *tmp = g_strjoinv (",", (gchar **) langs);
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "failed to get splash file for %s in %s",
+			     tmp, localedir);
+		return NULL;
+	}
+
+	/* decompress data */
+	stream_compressed = g_memory_input_stream_new_from_bytes (compressed_data);
+	conv = G_CONVERTER (g_zlib_decompressor_new (G_ZLIB_COMPRESSOR_FORMAT_GZIP));
+	stream_raw = g_converter_input_stream_new (stream_compressed, conv);
+	buf = g_malloc0 (buf_sz);
+	while ((len = g_input_stream_read (stream_raw,
+					   buf + buf_idx,
+					   buf_sz - buf_idx,
+					   NULL, error)) > 0) {
+		buf_idx += len;
+		if (buf_sz - buf_idx < chunk_size) {
+			buf_sz += chunk_size;
+			buf = g_realloc (buf, buf_sz);
+		}
+	}
+	if (len < 0) {
+		g_prefix_error (error, "failed to decompress file: ");
+		return NULL;
+	}
+	g_debug ("decompressed image to %" G_GSIZE_FORMAT "kb", buf_idx / 1024);
+	return g_bytes_new_take (g_steal_pointer (&buf), buf_idx);
+}
+
+static gboolean
+fu_plugin_uefi_update_splash (GError **error)
+{
+	fwup_resource *re = NULL;
+	guint best_idx = G_MAXUINT;
+	guint32 lowest_border_pixels = G_MAXUINT;
+#ifdef HAVE_FWUPD_GET_BGRT_INFO
+	int rc;
+#endif
+	guint32 screen_height = 640;
+	guint32 screen_width = 480;
+	g_autoptr(fwup_resource_iter) iter = NULL;
+	g_autoptr(GBytes) image_bmp = NULL;
+	struct {
+		guint32	 width;
+		guint32	 height;
+	} sizes[] = {
+		{ 640, 480 },
+		{ 800, 600 },
+		{ 1024, 768 },
+		{ 1920, 1080 },
+		{ 3840, 2160 },
+		{ 5120, 2880 },
+		{ 5688, 3200 },
+		{ 7680, 4320 },
+		{ 0, 0 }
```

Comment:
A comment about where these sizes come from would be good.  They look rather arbitrary.

Reply:
Fixed.